torch/csrc/autograd/engine.cpp

```diff
@@ -51,13 +52,19 @@ struct FunctionTask {
     , inputs(std::move(inputs)) {}
 };
 
+struct CompareFunctionTaskTime {
+  bool operator()(FunctionTask const & t1, FunctionTask const & t2) {
+    return t1.fn->time < t2.fn->time;
+  }
+};
+
 struct ReadyQueue {
-  std::deque<FunctionTask> queue;
+  std::priority_queue<FunctionTask, std::vector<FunctionTask>, CompareFunctionTaskTime> heap;
```

Comment:
If I understand correctly, `priority_queue` contains the smallest element on the top (if you use a comparator like the one you did), so the Engine will actually process the tasks in a DFS-like manner, where it always picks the earliest one. I think we wanted to do the converse - always pick the oldest one (to reproduce the steps in a reversed order).

Reply:
It's actually a max-heap. Quoting from cppreference:
```
Note that the Compare parameter is defined such that it returns true if its first argument comes before its second argument in a weak ordering. But because the priority queue outputs largest elements first, the elements that "come before" are actually output last. That is, the front of the queue contains the "last" element according to the weak ordering imposed by Compare.
```

Reply:
And empirically it behaves like the one we discussed. (see the last model in benchmark results.)